From 5f03c5f5cd008534a3145fa20cd8e145ecda8a74 Mon Sep 17 00:00:00 2001
From: Paul Naoumenko <paulx.naoumenko@intel.com>
Date: Mon, 14 Oct 2013 11:33:10 +0200
Subject: [PORT FROM MAIN] wifi: add support for bigcore platforms

BZ: 116022

This patch adds support in wpa_supplicant for bigcore
platforms.
1. It adds driver command for nl80211 interface for bigcore (non BCM)
   solutions
2. Removes BCM dependencies for bigcore

The driver command function for nl80211 has been ported
from OTC tree which supports bigcore platforms.

Change-Id: Ida2e9c1a4b97bbee6d2d872e6a0ca16dc88e4fe5
Orig-Change-Id: I6f9562a9bb75359f392f1fed1e2a1452366edfbd
Category: device enablement
Domain: CWS.WIFI-Bigcore
Origin: upstream
Upstream-Candidate: no,proprietary
Signed-off-by: Huzefa Kankroliwala <huzefa.nomanx.kankroliwala@intel.com>
---
 Android.mk                   |    4 +++
 src/drivers/driver_nl80211.c |   51 ++++++++++++++++++++++++++++++++++++++++++
 wpa_supplicant/Android.mk    |    4 +++
 3 files changed, 59 insertions(+), 0 deletions(-)

diff --git a/Android.mk b/Android.mk
index 4a74b24..72bbcd3 100644
--- a/Android.mk
+++ b/Android.mk
@@ -14,3 +14,7 @@ endif
 ifeq ($(WPA_SUPPLICANT_VERSION),VER_2_1_DEVEL)
 include $(call all-subdir-makefiles)
 endif
+
+ifeq ($(strip $(BOARD_WLAN_DEVICE)),wifi-bigcore)
+include $(LOCAL_DIR)/wpa_supplicant/Android.mk
+endif
diff --git a/src/drivers/driver_nl80211.c b/src/drivers/driver_nl80211.c
index c61740e..3434647 100644
--- a/src/drivers/driver_nl80211.c
+++ b/src/drivers/driver_nl80211.c
@@ -7871,6 +7871,54 @@ nla_put_failure:
 
 }
 
+#if defined(ANDROID) && defined(ANDROID_BIGCORE)
+static int nl80211_priv_driver_cmd(void *priv, char *cmd, char *buf, size_t buf_len)
+{
+	struct i802_bss *bss = priv;
+	struct wpa_driver_nl80211_data *drv = bss->drv;
+	struct wpa_signal_info sig;
+	int ret = 0, flags, rssi, linkspeed;
+	u8 macaddr[ETH_ALEN] = {};
+
+	if (os_strcasecmp(cmd, "MACADDR") == 0) {
+		ret = linux_get_ifhwaddr(drv->global->ioctl_sock, bss->ifname, macaddr);
+		if (!ret) {
+			ret = os_snprintf(buf, buf_len,
+					  "Macaddr = " MACSTR "\n", MAC2STR(macaddr));
+		}
+	} else if ((os_strcasecmp(cmd, "RSSI") == 0) ||
+		   (os_strcasecmp(cmd, "RSSI-APPROX") == 0)) {
+		if (!drv->associated) {
+			ret = -1;
+			goto priv_cmd_end;
+		}
+
+		ret = nl80211_get_link_signal(drv, &sig);
+		if (ret == 0) {
+			rssi = sig.current_signal;
+			ret = os_snprintf(buf, buf_len, "%s rssi %d\n", drv->ssid, rssi);
+		}
+	} else if (os_strcasecmp(cmd, "LINKSPEED") == 0) {
+		if (!drv->associated) {
+			ret = -1;
+			goto priv_cmd_end;
+		}
+
+		ret = nl80211_get_link_signal(drv, &sig);
+		if (ret == 0) {
+			linkspeed = sig.current_txrate / 1000;
+			ret = os_snprintf(buf, buf_len, "LinkSpeed %d\n", linkspeed);
+		}
+	} else {
+		wpa_printf(MSG_ERROR, "Unsupported command: %s", cmd);
+		ret = -1;
+	}
+
+priv_cmd_end:
+	return ret;
+}
+#endif /* ANDROID */
+
 
 static int wpa_driver_nl80211_connect(
 	struct wpa_driver_nl80211_data *drv,
@@ -10840,4 +10888,7 @@ const struct wpa_driver_ops wpa_driver_nl80211_ops = {
 #if defined(ANDROID) && !defined(NO_PRIVATE_LIB)
 	.driver_cmd = wpa_driver_nl80211_driver_cmd,
 #endif
+#if defined(ANDROID) && defined(ANDROID_BIGCORE)
+	.driver_cmd = nl80211_priv_driver_cmd,
+#endif /* ANDROID */
 };
diff --git a/wpa_supplicant/Android.mk b/wpa_supplicant/Android.mk
index b8e90d2..88f301b 100644
--- a/wpa_supplicant/Android.mk
+++ b/wpa_supplicant/Android.mk
@@ -34,6 +34,10 @@ L_CFLAGS += -DANDROID_P2P
 L_CFLAGS += -DP2P_CONCURRENT_SEARCH_DELAY=0
 endif
 
+ifeq ($(BOARD_WLAN_DEVICE), wifi-bigcore)
+L_CFLAGS += -DANDROID_BIGCORE
+endif
+
 ifeq ($(BOARD_WLAN_DEVICE), qcwcn)
 L_CFLAGS += -DANDROID_P2P
 endif
-- 
1.7.4.1

